agent/spec/lib/kontena/observer_spec.rb

```diff
@@ -1,79 +1,198 @@
+  def observed_value
+    @observed_value
+  end
+  def ordered?
+    @ordered
+  end
 
-      attr_reader :state, :values
+  def test_exclusive_observe(*observables, **options)
+    exclusive {
+      return observe(*observables, **options)
+    }
+  end
 
-      def initialize(*observables)
-        @state = observe(*observables) do |*values|
-          @values = values
-        end
-      end
+  def crash
+    fail
+  end
+end
 
-      def ready?
-        @state.ready?
-      end
+class TestObserverStandalone
+  include Kontena::Observer
+end
 
-      def crash
-        fail
-      end
-    end
-  end
+describe Kontena::Observer, :celluloid => true do
+  let(:observable_class) { TestObservable }
+  let(:observer_class) { TestObserver }
+
+  subject { observer_class.new() }
 
   it "raises synchronously if given an invalid actor", :celluloid => true, :log_celluloid_actor_crashes => false do
-    expect{observer_class.new('foo')}.to raise_error(NoMethodError, /undefined method `add_observer' for "foo":String/)
+    expect{subject.test_observe_async('foo')}.to raise_error(NoMethodError, /undefined method `add_observer' for "foo":String/)
   end
 
   context "For a single observable" do
     let(:observable) { observable_class.new }
 
-    subject { observer_class.new(observable) }
-
     let(:object) { double(:test) }
 
-    it "does not observe any value if not yet updated", :celluloid => true do
-      expect(subject).to_not be_ready
-    end
+    describe '#observe_async' do
+      it "does not observe any value if not yet updated" do
+        subject.test_observe_async(observable)
 
-    it "immediately yields an updated value", :celluloid => true do
-      observable.update_observable object
+        expect(subject).to_not be_observed
+      end
 
-      subject
+      it "immediately yields an updated value" do
+        observable.update_observable object
 
-      expect(subject).to be_ready
-      expect(subject.values).to eq [object]
-    end
+        subject.test_observe_async(observable)
 
-    it "later yields after updating value", :celluloid => true do
-      subject
+        expect(subject).to be_observed
+        expect(subject.observed_values).to eq [object]
+      end
+
+      it "later yields after updating value" do
+        subject.test_observe_async(observable)
 
-      expect(subject).to_not be_ready
+        expect(subject).to_not be_observed
 
-      observable.update_observable object
+        observable.update_observable object
 
-      expect(subject).to be_ready
-      expect(subject.values).to eq [object]
+        expect(subject).to be_observed
+        expect(subject.observed_values).to eq [object]
+      end
+
+      it "crashes if the observable does", :log_celluloid_actor_crashes => true do
+        subject.test_observe_async(observable)
+
+        expect{observable.crash}.to raise_error(RuntimeError)
+
+        expect{subject.ping; subject.observed?}.to raise_error(Celluloid::DeadActorError)
```

Comment:
Fixed, the `oservable.crash` [`Celluloid::Call::Sync#dispatch`](https://github.com/celluloid/celluloid/blob/v0.17.3/lib/celluloid/call/sync.rb#L21) was sending the error response back to the calling rspec thread before raising within the observable actor thread, and having that send out the `ExitEvent` messages to the linked actors.

It was thus possible for the rspec thread to see the crash call response, and send the ping calls to the observer actor before the observable actor propagated the crash.

```
D, [2017-08-22T17:18:20.769611 #32054] DEBUG -- TestObservable: observer: Observer<TestObserver>...
D, [2017-08-22T17:18:20.769870 #32054] DEBUG -- TestObserver: observe async Observable<TestObservable>...
D, [2017-08-22T17:18:20.769926 #32054] DEBUG -- TestObserver: observe async Observable<!TestObservable>...
D, [2017-08-22T17:18:20.772056 #32054] DEBUG -- TestObserver: ping
D, [2017-08-22T17:18:20.772259 #32054] DEBUG -- TestObserver: ping
D, [2017-08-22T17:18:20.772418 #32054] DEBUG -- TestObserver: ping
E, [2017-08-22T17:18:20.771622 #32054] ERROR -- : Actor crashed!
RuntimeError: 
	/home/kontena/kontena/kontena/agent/spec/lib/kontena/observable_spec.rb:11:in `crash'
	/home/kontena/kontena/kontena/agent/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
	/home/kontena/kontena/kontena/agent/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
	/home/kontena/kontena/kontena/agent/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:16:in `dispatch'
	/home/kontena/kontena/kontena/agent/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
	/home/kontena/kontena/kontena/agent/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
	/home/kontena/kontena/kontena/agent/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
	/home/kontena/kontena/kontena/agent/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
	/home/kontena/kontena/kontena/agent/vendor/bundle/ruby/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
E, [2017-08-22T17:18:20.772906 #32054] ERROR -- : Actor crashed!
```